### ISSUE TITLE
Remove a cell from evaluation queue

### ISSUE THREAD

**Issue by cossio** [jupyterlab/jupyterlab#7825]
## Feature request

As far as I can tell, in order to remove one cell from the evaluation queue, the only option is to completely interrupt the kernel (stopping all evaluations, not just the cell of interest).

It would be nice to have the ability to interrupt computation only on the currently running cell (and automatically moving on to the next cell in the evaluation queue), and also to be able to remove a selected cell (that might might be waiting for other cells to run) from the evaluation queue.

In Mathematica it is possible to do this.

**Comment by stefansimik** [comment]
That would be veery practical.
It happens so frequently, that I run everything, but then I see 
I would like to remove some (very slow but not important) cell 
from evaluation queue to speedup displaying results.

**Comment by shugybugy-assaf** [comment]
It has to be implemented !!!
DataBricks notebooks have this super useful feature for a reason !

**Comment by shugybugy-assaf** [comment]
Duplicate of a feature request from 2017
https://github.com/jupyter/notebook/issues/2340

**Comment by ismael-elatifi** [comment]
As I explained [here](https://github.com/jupyter/notebook/issues/2340#issuecomment-732164654) , a simple way to achieve this would be to add a mode to send a cell to kernel for execution only when previous cell execution is done.
With this new done, as long as previous cell are still running we can easily cancel execution of subsequent cells because there are not yet sent to the kernel.

**Comment by cossio** [comment]
Any updates on this?

**Comment by blois** [comment]
@ismael-elatifi - Google Colab used to have that execution model but we moved away from it because:
1) It slowed down run-all of a lot of small cells.
2) It would pause execution for users who run-all then close their laptop.

Jason's comment at https://github.com/jupyter/notebook/issues/2340#issuecomment-586311279 sounds like a more robust method.

**Comment by jasongrout** [comment]
I was brainstorming ideas about how to expose the execution queue to users.

The kernel execution queue can be populated by any connected client, so it's not enough to show just cells that I've sent to the kernel. Also, a notebook cell can be edited after it was sent to the kernel, so it is not enough to even just indicate in a notebook that a cell was sent (i.e., the cell that is visible may not match with what code was sent).

How about:
1. When a kernel receives an execute request, it immediately broadcasts a message (perhaps the execute_input) informing clients that a new thing has been added to the execution queue. The broadcast message should include the header of the original execution request so that we get a unique identifier (the msg id) and a sense of the source (the client id).
2. Clicking on the kernel execution status icon shows a dropdown of all current pending execution requests in the kernel, perhaps the first nonblank line of the input, with an option to expand to show all input. For execution requests that we sent, perhaps a link to take us to the relevant notebook or console cell. Each execution request has an option to cancel just it, or cancel it and all following requests. Cancellation can maybe use the interrupt request idea from https://github.com/jupyter/notebook/issues/2340#issuecomment-586311279, i.e., change the interrupt request message to include a list of msg ids that should be cancelled in the queue.
3. Perhaps change something about a pending cell when it is modified from the state that was sent, so it is clear that it has been modified. Also, perhaps include a link from the cell to open this read-only execution list to the appropriate place.
4. There needs to be some method (probably a debug method on the control channel?) for requesting the current execution queue for clients that connect to a busy kernel.

**Comment by ismael-elatifi** [comment]
> I was brainstorming ideas about how to expose the execution queue to users.
> 
> The kernel execution queue can be populated by any connected client, so it's not enough to show just cells that I've sent to the kernel. Also, a notebook cell can be edited after it was sent to the kernel, so it is not enough to even just indicate in a notebook that a cell was sent (i.e., the cell that is visible may not match with what code was sent).
> 
> How about:
> 
> 1. When a kernel receives an execute request, it immediately broadcasts a message (perhaps the execute_input) informing clients that a new thing has been added to the execution queue. The broadcast message should include the header of the original execution request so that we get a unique identifier (the msg id) and a sense of the source (the client id).
> 2. Clicking on the kernel execution status icon shows a dropdown of all current pending execution requests in the kernel, perhaps the first nonblank line of the input, with an option to expand to show all input. For execution requests that we sent, perhaps a link to take us to the relevant notebook or console cell. Each execution request has an option to cancel just it, or cancel it and all following requests. Cancellation can maybe use the interrupt request idea from [Interrupt planned cell execution jupyter/notebook#2340 (comment)](https://github.com/jupyter/notebook/issues/2340#issuecomment-586311279), i.e., change the interrupt request message to include a list of msg ids that should be cancelled in the queue.
> 3. Perhaps change something about a pending cell when it is modified from the state that was sent, so it is clear that it has been modified. Also, perhaps include a link from the cell to open this read-only execution list to the appropriate place.
> 4. There needs to be some method (probably a debug method on the control channel?) for requesting the current execution queue for clients that connect to a busy kernel.

This seems a bit complicated. Something much simpler and alrerady very useful would be to add buttons `Remove cell from queue` and `Remove cell and cells below from queue`.
The UI would display an indicator for which the removing from the queue was successfull and a message for cells that could not be removed from the queue (for example because already running).

**Comment by jasongrout** [comment]
> This seems a bit complicated. Something much simpler and already very useful would be to **add just one button `Remove cell and cells below from queue`**.

Good point - in the common simpler case of a single user running a single notebook connected to a kernel, supporting that workflow sounds great. Right now this could go on the right-click menu, and eventually in a cell toolbar when we have a cell-specific toolbar? Or perhaps it goes in the notebook toolbar (perhaps beside the interrupt button?)

I'm not sure how it works now if you edit a pending cell and resubmit. Does it let you submit a new execution request? If so, then canceling execution needs to differentiate between the two execute requests.

Regardless, the first step here would be pushing forward a protocol change that allows us to cancel a pending execution.

**Comment by Alfred-Mountfield** [comment]
> **`Remove cell and cells below from queue`**

Not trying to be pedantic, I just want to check I understand the proposal. When you say "cells below from queue" are you referring to position within a notebook, or within the queue? In the context of running all cells in a notebook linearly and actually wanting to pause half way down then they should be equivalent. 

I'd definitely want the ability to remove a single cell, rather than always removing all cells below it from the queue. Especially if I have just accidentally queued a cell higher up than the ones currently executing.

**Comment by Alfred-Mountfield** [comment]
Also I wrote all of this out without properly understanding the discussion above, so I ended up rehashing some of the existing conversation but I hope phrasing it differently and giving a concrete example can help with future discussion if a decision is made to move past the simple use-case.

I actually just encountered this issue while looking up a solution for a different use-case though, where I ran-all, am most of the way through an expensive notebook, I then modified cell 3 and queued it up again, and I realise I actually want to cancel cell 3 (while I'm almost done executing a later cell, lets say 7).

So after I ran-all, at one point I was in a state something like this
Finished Cells = `[0, 1, 2, 3, 4, 5, 6]`
Queued Cells = `[7, 8]`

I then edited cell 3 and queued it
Finished Cells = `[0, 1, 2, 3, 4, 5, 6]`
Queued Cells = `[7, 8, 3(modified)]`

And now I'm unable to queue any other workloads, as I can't remove `3(modified)` from the queue. So now I have to either cancel `[7, 8, 3(modified)]` and restart `[7, 8]` (which is only possible because they're pure functions), or I have to wait for `[7, 8]` to finish (and there's no process to notify me when that's the case, so I have to keep checking) and then interrupt `3(modified)` and start queuing new things.

> Clicking on the kernel execution status icon shows a dropdown of all current pending execution requests in the kernel, perhaps the first nonblank line of the input, with an option to expand to show all input. For execution requests that we sent, perhaps a link to take us to the relevant notebook or console cell. Each execution request has an option to cancel just it, or cancel it and all following requests. Cancellation can maybe use the interrupt request idea from https://github.com/jupyter/notebook/issues/2340#issuecomment-586311279, i.e., change the interrupt request message to include a list of msg ids that should be cancelled in the queue.

I think it becomes particularly difficult from a UX perspective if the user is adding/removing cells and changing indices, and changing the contents of cells that are currently being evaluated. It means we can't easily give indications of where the cells are being executed. So rather than linking to the cell or taking the user to the correct notebook, perhaps it could just include (in the message) a snapshot of the contents of the cell at the time where it was queued and expose that?

**Comment by ismael-elatifi** [comment]
> > **`Remove cell and cells below from queue`**
> 
> Not trying to be pedantic, I just want to check I understand the proposal. When you say "cells below from queue" are you referring to position within a notebook, or within the queue? In the context of running all cells in a notebook linearly and actually wanting to pause half way down then they should be equivalent.
> 
> I'd definitely want the ability to remove a single cell, rather than always removing all cells below it from the queue. Especially if I have just accidentally queued a cell higher up than the ones currently executing.

Yes indeed having the possibility to remove a single cell from the queue would be even better but it is more tricky and possibly not safe if cells after in the queue depended on the cell removed.
Removing the cell and all cells after in the queue is easier and safe that is why I proposed that as a first concrete solution.
If you have this `Remove cell and cells below from queue` button you can also remove a single cell by removing from cell X and rerun from cell X+1.

**Comment by dennysemko** [comment]
I'm surprised this very useful feature has not yet been implemented, this could've saved a lot of time for me over the years.

**Comment by filippo1335** [comment]
still need this, even for resolving bugs where cell is awaiting in the queue indefinitely

**Comment by eggrollofchaos** [comment]
Need this as well! +1

**Comment by krassowski** [comment]
If there was an evaluation queue UI in the sidebar, we would not need a separate "scroll to the currently executing cell" button (https://github.com/jupyterlab/jupyterlab/pull/16880), we would just click on the cell in the queue.
